staging/src/k8s.io/cloud-provider/fake/fake.go

```diff
@@ -222,13 +224,21 @@ func (f *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, serv
 	return status, f.Err
 }
 
+func (f *Cloud) markUpdateCall(service *v1.Service, nodes []*v1.Node) {
+	f.Lock.Lock()
+	defer f.Lock.Unlock()
+	update := UpdateBalancerCall{service, nodes}
+	f.UpdateCalls = append(f.UpdateCalls, update)
+	if f.UpdateCallCb != nil {
+		f.UpdateCallCb(update)
+	}
+}
+
 // UpdateLoadBalancer is a test-spy implementation of LoadBalancer.UpdateLoadBalancer.
 // It adds an entry "update" into the internal method call record.
 func (f *Cloud) UpdateLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) error {
+	f.markUpdateCall(service, nodes)
```

Comment:
this was running after `addCall`

Reply:
Yes, but if someone specifies a `RequestDelay` what we want is to register the call and then wait `RequestDelay`. The existing tests never utilised `RequestDelay`, so it didn't matter to them that we registered the call after having waited - but for the test case I've added, it does.

Reply:
>  he existing tests never utilised RequestDelay, so it didn't matter to them that we registered

but is not about the existing tests, future tests may want to use it and it will not work, this is not correct to modify a behavior like this. Since our test does not use RequestDelay it does not matter for us the order, so we should keep it

Reply:
>  future tests may want to use it and it will not work

What will not work? 

> Since our test does not use RequestDelay it does not matter for us the order, so we should keep it

I don't understand: what is "our test"?

Reply:
> What will not work?

this change runs `markUpdateCall` before `RequestDelay` , this test does not set `RequestDelay`, why do you have to change the order? am I missing something?

Reply:
The test I've added (`TestSlowNodeSync`) does set `RequestDelay`. The reason `markUpdateCall` needs to run before waiting `RequestDelay`, is because we need to mock the behaviour whereby we trigger a service sync while the node sync is processing. Essentially the test does the following:

1. Process a node sync. 
2. [node sync] `UpdateLoadBalancer` will be called for service1 (takes `RequestDelay` to process)
3. node3 is created 
4. service sync is triggered for service2
5. [service sync] `EnsureLoadBalancer` will be called for service2
6. [node sync] `UpdateLoadBalancer` will be called for service2 

If we run `markUpdateCall` after waiting `RequestDelay` we can't consistently mock the behaviour and might have the following instead:

1. We process a node sync. 
2. [node sync] `UpdateLoadBalancer` will be called for service1 
3. [node sync] `UpdateLoadBalancer` will be called for service2 
4. node3 is created 
5. service sync is triggered for service2
6. [service sync] `EnsureLoadBalancer` will be called for service2 

i.e: we need to create node3 and trigger the service sync for service2 _while_ the we are processing service1 from the node sync. This validates that we don't remove nodes from the service2 LB set, because the node sync has stale node data. This is the bug this PR is trying to fix.    

For all existing tests, this behaviour is never tested, so it doesn't matter if `markUpdateCall` is called before waiting `RequestDelay` or after.

Reply:
can't you add an updateCall that adds the delay instead of having to modify the behavior?

Reply:
Done